lib/spack/spack/spec.py

```diff
@@ -52,8 +53,10 @@
   spec-list    = { spec [ dep-list ] }
   dep_list     = { ^ spec }
   spec         = id [ options ]
-  options      = { @version-list | +variant | -variant | ~variant |
-                   %compiler | arch=architecture | [ flag ]=value}
+  options      = { @version-list | ++variant | +variant |
+                   --variant | -variant | ~~variant | ~variant |
+                   %compiler | arch=architecture |
+                   [ flag ]==value | [ flag ]=value}
```

Comment:
Side note: This grammar is incorrect (prior to this PR) because it doesn't account for variants specified by `name=value`.